tools/keycloak/sync/SyncFactory.cs

```diff
@@ -67,34 +71,62 @@ public async Task<int> SyncAsync()
         }
 
         #region Helpers
+
         /// <summary>
         /// Activate the service account.
         /// </summary>
         /// <returns></returns>
         private async Task ActivateAccountAsync()
         {
-            var aRes = await _client.SendAsync($"{_options.Api.Uri}/auth/activate", HttpMethod.Post);
-            if (!aRes.IsSuccessStatusCode) throw new HttpClientRequestException(aRes);
+            var aRes = await _pimsClient.SendAsync("auth/activate", HttpMethod.Post);
+            if (!aRes.IsSuccessStatusCode)
+            {
+                throw new HttpClientRequestException(aRes);
+            }
         }
 
         #region Claim/Role
+
         /// <summary>
         /// Fetch all Claims in PIMS and link or create them in keycloak.
         /// </summary>
         /// <param name="log"></param>
+        /// <param name="errorLog"></param>
         /// <returns></returns>
-        private async Task SyncClaimsAsync(StringBuilder log)
+        private async Task SyncClaimsAsync(StringBuilder log, StringBuilder errorLog)
         {
-            var claims = await _client.HandleRequestAsync<Api.Models.PageModel<PModel.ClaimModel>>(HttpMethod.Get, $"{_options.Api.Uri}/admin/claims?page=1&quantity=100"); // TODO: Handle paging.
-            foreach (var claim in claims?.Items)
+            IEnumerable<PModel.ClaimModel> claims = new List<PModel.ClaimModel>();
+            Api.Models.PageModel<PModel.ClaimModel> claimsPage = null;
+            int page = 1;
+            do
             {
-                var krole = await GetKeycloakRoleAsync(claim);
-                claim.Key = krole.Id;
-                claim.KeycloakRoleId = krole.Id;
+                claimsPage = await _pimsClient.HandleRequestAsync<Api.Models.PageModel<PModel.ClaimModel>>(HttpMethod.Get, $"admin/claims?page={page++}&quantity=50");
+                claims = claims.Concat(claimsPage.Items);
+            }
+            while (claimsPage != null && claimsPage.Items.Any());
+
+            var allKeycloakRoles = await _keycloakManagementClient.HandleRequestAsync<ResponseWrapperModel<RoleModel[]>>(HttpMethod.Get, $"{_keycloakManagementClient.GetIntegrationEnvUri()}/roles");
+            var keycloakRoles = allKeycloakRoles.Data.Where(x => x.Composite.HasValue && x.Composite.Value == false);
```

Comment:
q: If it wants to get roles should the Composite.Value be true instead if it is not null?

Reply:
Hmm, I'm worried I'm not understanding, this is checking for Composite.Value to be false and not null.

Maybe I can provide a description of composites and roles in PSP vs keycloak.

So keycloak doesn't support claims or roles in the same way that we do. They just support roles. Logically, a keycloak role is most similar to our claims, as it is basically just a string. If the user has that string, they have that claim. Where it gets confusing is that a keycloak role may also be a composite of some other keycloak roles. This is close to being equivalent to our roles in PSP. So what I'm doing here is:
if a keycloak role exists, and it has at least one composite child it should be treated as a PSP role.
If a keycloak role exists and it doesn't have any composite children it should be treated as a PSP claim.

Any ideas on how to make the separation between the two or the naming more robust are welcome.